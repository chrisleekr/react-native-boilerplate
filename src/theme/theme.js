import {Dimensions} from 'react-native';
import {colors} from './colors';
const {width, height} = Dimensions.get('window');

// Themes:- Can alter values here. Can only be consumed through Context (see useTheme.js file)
const themes = {
  light: {
    name: 'light',
    barStyle: 'light-content',
    color: '#000000',
    labelColor: '#525E75',
    invertColor: '#ffffff',
    layoutBg: '#f0f0f0',
    headerBg: '#f6f7f8',
    cardBg: '#e4dfa7',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#B00020',
    buttonActiveColor: '#6200EE',
    buttonDisabledColor: 'rgba(0, 0, 0, 0.26)',
    card: {
      shadowColor: '#000000',
      borderColor: '#B00020'
    },
    ...colors
  },
  dark: {
    name: 'dark',
    barStyle: 'dark-content',
    color: '#ffffff',
    labelColor: '#F1DDBF',
    invertColor: '#000000',
    layoutBg: '#121212',
    headerBg: '#1e1e1e',
    cardBg: '#3f4045',
    accent: '#0071ff',
    error: '#B00020',
    errorText: '#ed4337',
    lightBg: '#F5EE9E',
    borderColor: '#ffffff',
    buttonActiveColor: '#BB86FC',
    buttonDisabledColor: 'rgba(255,255,255, 0.26)',
    card: {
      shadowColor: '#121212',
      borderColor: '#ffffff'
    },
    ...colors
  }
};

const window = {
  // app dimensions
  width,
  height
};

export {themes, window};
