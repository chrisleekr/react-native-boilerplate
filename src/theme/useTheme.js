/**
 * For Dark/Light Mode
 * @exports ThemeContext, ThemeProvider, useTheme
 */

import React, {useState, createContext, useContext} from 'react';
import {Appearance} from 'react-native';

// Import preconfigured themes from theme file
import {themes} from './theme';

// Context
export const ThemeContext = createContext(null);

// Provider to be used in index/App/or top of any parent
export const ThemeProvider = ({children}) => {
  const colorScheme = Appearance.getColorScheme();

  const [theme, setTheme] = useState(themes[colorScheme]);
  return (
    <ThemeContext.Provider value={{theme, setTheme}}>
      {children}
    </ThemeContext.Provider>
  );
};

// useTheme hook for easy access to theme and setTheme
export const useTheme = () => {
  const {theme, setTheme} = useContext(ThemeContext);

  const toggleTheme = v => {
    setTheme(v ? themes.dark : themes.light);
  };

  return {theme, toggleTheme};
};
