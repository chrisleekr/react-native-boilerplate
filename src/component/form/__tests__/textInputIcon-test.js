import 'react-native';
import React from 'react';

import {
  cleanup,
  fireEvent,
  render,
  within
} from '@testing-library/react-native';

import {ThemeProvider} from '../../../theme/useTheme';

import TextInputIcon from '../textInputIcon';

let rendered;
let component;
let element;

const onPressIcon = jest.fn();
const onBlur = jest.fn();
const onChangeTodo = jest.fn();
const setFieldTouched = jest.fn();

describe('TextInputIcon', () => {
  afterEach(cleanup);

  beforeEach(() => {
    rendered = render(
      <ThemeProvider>
        <TextInputIcon
          label={'My Label'}
          containerStyle={{
            backgroundColor: '#000000'
          }}
          textInputStyle={{
            fontSize: 15
          }}
          iconName="plus"
          iconSize={15}
          iconStyle={{
            marginLeft: 10
          }}
          onPressIcon={onPressIcon}
          isLoading={false}
          field={{
            name: 'todo',
            onBlur,
            onChange: jest.fn().mockImplementation(name => {
              switch (name) {
                case 'todo':
                  return onChangeTodo;
              }
            }),
            value: 'Initial text'
          }}
          form={{
            errors: {},
            touched: {},
            setFieldTouched
          }}
          keyboardType="number-pad"
        />
      </ThemeProvider>
    );

    component = rendered.getByTestId('text-input-icon');
  });

  it('matches snapshot', () => {
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  describe('onChangeText', () => {
    beforeEach(() => {
      element = within(component).getByTestId('input');

      fireEvent(element, 'onChangeText', 'New Text');
    });

    it('should trigger onChangeTodo', () => {
      expect(onChangeTodo).toHaveBeenCalledWith('New Text');
    });
  });

  describe('onBlur', () => {
    beforeEach(() => {
      element = within(component).getByTestId('input');

      fireEvent(element, 'onBlur');
    });

    it('should trigger setFieldTouched', () => {
      expect(setFieldTouched).toHaveBeenCalledWith('todo');
    });

    it('should trigger onBlur', () => {
      expect(onBlur).toHaveBeenCalledWith('todo');
    });
  });

  describe('onPressIcon', () => {
    beforeEach(() => {
      element = within(component).getByTestId('icon');

      fireEvent(element, 'onPress');
    });

    it('should trigger onPressIcon', () => {
      expect(onPressIcon).toHaveBeenCalled();
    });
  });

  describe('isLoading', () => {
    beforeEach(() => {
      rendered = render(
        <ThemeProvider>
          <TextInputIcon
            label={'My Label'}
            containerStyle={{
              backgroundColor: '#000000'
            }}
            textInputStyle={{
              fontSize: 15
            }}
            iconName="plus"
            iconSize={15}
            iconStyle={{
              marginLeft: 10
            }}
            onPressIcon={onPressIcon}
            isLoading={true}
            field={{
              name: 'todo',
              onBlur,
              onChange: jest.fn().mockImplementation(name => {
                switch (name) {
                  case 'todo':
                    return onChangeTodo;
                }
              }),
              value: 'Initial text'
            }}
            form={{
              errors: {},
              touched: {},
              setFieldTouched
            }}
            keyboardType="number-pad"
          />
        </ThemeProvider>
      );

      component = rendered.getByTestId('text-input-icon');
    });

    it('matches snapshot', () => {
      expect(rendered.toJSON()).toMatchSnapshot();
    });

    it('shows activity indicator', () => {
      expect(
        within(component).queryByTestId('activity-indicator')
      ).toBeTruthy();
    });

    it('does not show icon', () => {
      expect(within(component).queryByTestId('icon')).toBeFalsy();
    });
  });

  describe('hasError', () => {
    beforeEach(() => {
      rendered = render(
        <ThemeProvider>
          <TextInputIcon
            label={'My Label'}
            containerStyle={{
              backgroundColor: '#000000'
            }}
            textInputStyle={{
              fontSize: 15
            }}
            iconName="plus"
            iconSize={15}
            iconStyle={{
              marginLeft: 10
            }}
            onPressIcon={onPressIcon}
            isLoading={true}
            field={{
              name: 'todo',
              onBlur,
              onChange: jest.fn().mockImplementation(name => {
                switch (name) {
                  case 'todo':
                    return onChangeTodo;
                }
              }),
              value: 'Initial text'
            }}
            form={{
              errors: {
                todo: 'Please enter new todo.'
              },
              touched: {
                todo: true
              },
              setFieldTouched
            }}
            keyboardType="number-pad"
          />
        </ThemeProvider>
      );

      component = rendered.getByTestId('text-input-icon');
    });

    it('matches snapshot', () => {
      expect(rendered.toJSON()).toMatchSnapshot();
    });

    it('has error style', () => {
      expect(
        within(component).queryByTestId('input').props.style[2]
      ).toStrictEqual({
        borderColor: 'red',
        borderWidth: 1
      });
    });
  });
});
