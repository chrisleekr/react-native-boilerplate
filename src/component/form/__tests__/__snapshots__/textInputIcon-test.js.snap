// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputIcon hasError matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
      Object {
        "backgroundColor": "#000000",
      },
    ]
  }
  testID="text-input-icon"
>
  <TextInput
    keyboardType="number-pad"
    label="My Label"
    onBlur={[Function]}
    onChangeText={[Function]}
    style={
      Array [
        Object {
          "borderWidth": 1,
          "height": 35,
          "paddingBottom": 0,
          "paddingTop": 0,
          "textAlignVertical": "center",
          "width": "100%",
        },
        Object {
          "fontSize": 15,
        },
        Object {
          "borderColor": "red",
          "borderWidth": 1,
        },
        Object {
          "backgroundColor": "#e4dfa7",
        },
      ]
    }
    testID="input"
    value="Initial text"
  />
  <View
    style={
      Array [
        Object {
          "position": "absolute",
          "right": 5,
          "top": 8,
          "zIndex": 1,
        },
      ]
    }
    testID="activity-indicator"
  >
    <ActivityIndicator
      color="rgba(0, 0, 0, 0.26)"
    />
  </View>
</View>
`;

exports[`TextInputIcon isLoading matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
      Object {
        "backgroundColor": "#000000",
      },
    ]
  }
  testID="text-input-icon"
>
  <TextInput
    keyboardType="number-pad"
    label="My Label"
    onBlur={[Function]}
    onChangeText={[Function]}
    style={
      Array [
        Object {
          "borderWidth": 1,
          "height": 35,
          "paddingBottom": 0,
          "paddingTop": 0,
          "textAlignVertical": "center",
          "width": "100%",
        },
        Object {
          "fontSize": 15,
        },
        undefined,
        Object {
          "backgroundColor": "#e4dfa7",
        },
      ]
    }
    testID="input"
    value="Initial text"
  />
  <View
    style={
      Array [
        Object {
          "position": "absolute",
          "right": 5,
          "top": 8,
          "zIndex": 1,
        },
      ]
    }
    testID="activity-indicator"
  >
    <ActivityIndicator
      color="rgba(0, 0, 0, 0.26)"
    />
  </View>
</View>
`;

exports[`TextInputIcon matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
      Object {
        "backgroundColor": "#000000",
      },
    ]
  }
  testID="text-input-icon"
>
  <TextInput
    keyboardType="number-pad"
    label="My Label"
    onBlur={[Function]}
    onChangeText={[Function]}
    style={
      Array [
        Object {
          "borderWidth": 1,
          "height": 35,
          "paddingBottom": 0,
          "paddingTop": 0,
          "textAlignVertical": "center",
          "width": "100%",
        },
        Object {
          "fontSize": 15,
        },
        undefined,
        Object {
          "backgroundColor": "#e4dfa7",
        },
      ]
    }
    testID="input"
    value="Initial text"
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginLeft": 10,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
        "textAlignVertical": "center",
        "top": 2,
        "zIndex": 1,
      }
    }
    testID="icon"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 15,
          },
          undefined,
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      Û∞êï
    </Text>
  </View>
</View>
`;
